src/tools/ceph_objectstore_tool.cc

```diff
@@ -680,59 +685,86 @@ int do_trim_pg_log(ObjectStore *store, const coll_t &coll,
 	continue;
       if (p->key().substr(0, 7) == string("missing"))
 	continue;
-      if (p->key().substr(0, 4) == string("dup_"))
-	continue;
-
       bufferlist bl = p->value();
       auto bp = bl.cbegin();
-      pg_log_entry_t e;
-      try {
-	e.decode_with_checksum(bp);
-      } catch (const buffer::error &e) {
-	cerr << "Error reading pg log entry: " << e.what() << std::endl;
-      }
-      if (debug) {
-	cerr << "read entry " << e << std::endl;
+
+      if (p->key().substr(0, 4) == string("dup_")) {  
+        pg_log_dup_t dup;
+	try {
+	  dup.decode(bp);
+	  pg_log.dups.push_back(dup);
+	  if (debug) {
+            cerr << "read entry " << dup << std::endl;
+          }
+	} catch (buffer::error& e) {
+	  cerr << "error reading log dup: " << p->key() << std::endl;
+	  return -EFAULT;
+	}
+      } else {
+        pg_log_entry_t e;
+        try {
+          e.decode_with_checksum(bp);
+          pg_log.log.push_back(e);
+	  if (debug) {
+            cerr << "read entry " << e << std::endl;
+          }
+	} catch (const buffer::error &e) {
+          cerr << "Error reading pg log entry: " << e.what() << std::endl;
+        }
       }
-      if (e.version.version > trim_to) {
-	done = true;
-	break;
+      // We have enough pg logs entries to trim
+      if ((pg_log.log.size() + pg_log.dups.size()) >= trim_at_once) {
+        done = true;
+        break;
       }
-      keys_to_trim.insert(p->key());
-      new_tail = e.version;
-      if (keys_to_trim.size() >= trim_at_once)
-	break;
     }
-
     if (!p->valid())
       done = true;
+    pg_log.index();
+
     } // deconstruct ObjectMapIterator
 
-    // delete the keys
-    if (!dry_run && !keys_to_trim.empty()) {
-      cout << "Removing keys " << *keys_to_trim.begin() << " - " << *keys_to_trim.rbegin() << std::endl;
-      ObjectStore::Transaction t;
-      t.omap_rmkeys(coll, oid, keys_to_trim);
-      store->queue_transaction(ch, std::move(t));
-      ch->flush();
+    if (!dry_run && trim_to > pg_log.tail) {
+      std::set<std::string> trimmed;
+      std::set<std::string> trimmed_dups;
+      eversion_t write_from_dups = eversion_t::max();
+
+      pg_log.trim(g_ceph_context, trim_to, &trimmed, &trimmed_dups, &write_from_dups);
+      // Is there any trimming to do?
+      if (!trimmed.empty() || !trimmed_dups.empty()) {
+        info.log_tail = pg_log.tail;
+        new_tail = true;
+        
+        ObjectStore::Transaction t;
+        if (!trimmed.empty()) {
+	  cout << "Removing keys " << *trimmed.begin() << " - " << *trimmed.rbegin() << std::endl;
```

Comment:
nit: if we are changin this, I would prefer messages like this to go to `cerr`, not `cout`. Because they are more like log messages. The same is for other messages below.